mailgun/route_handlers.py

```diff
@@ -30,193 +30,222 @@
         if message_id:
-            cache_key = settings.CACHE_KEY_MESSAGE_HANDLED_BY_MESSAGE_ID_AND_RECIPIENT % (message_id, recipient)
+            cache_key = (
+                settings.CACHE_KEY_MESSAGE_HANDLED_BY_MESSAGE_ID_AND_RECIPIENT
+                    % (message_id, recipient))
             if cache.get(cache_key):
                 logger.warning(u'Message-Id %s was posted to the route handler '
-                               u'for %s, but we\'ve already handled that.  Dropping.',
-                               recipient, message_id)
+                               u"for %s, but we've already handled that.  "
+                               u'Skipping.', recipient, message_id)
                 continue
 
-        # make sure the mailing list exists
-        bounce_back_email_template = None
         try:
-            ml = MailingList.objects.get_or_create_or_delete_mailing_list_by_address(recipient)
-        except MailingList.DoesNotExist:
+            _handle_recipient(request, recipient)
+        except JsonResponse as error_response:
+            return error_response
+
+    return JsonResponse({'success': True})
+
+
+def _handle_recipient(request, recipient):
+    '''
+    The logic behind whether an email will be forwarded to list members or
+    trigger a bounce email can be complicated.  A (hopefully simpler to follow)
+    matrix can be found on the wiki:
+        https://confluence.huit.harvard.edu/display/TLT/LTI+Emailer
+    '''
+    attachments, inlines = _get_attachments_inlines(request)
+    body_html = request.POST.get('body-html', '')
+    body_plain = request.POST.get('body-plain', '')
+    cc_list = address.parse_list(request.POST.get('Cc'))
+    message_id = request.POST.get('Message-Id')
+    sender = request.POST.get('sender')
+    subject = request.POST.get('subject')
+    to_list = address.parse_list(request.POST.get('To'))
+
+    logger.debug(u'Handling recipient %s, from %s, subject %s, message id %s',
+                 recipient, sender, subject, message_id)
+
+    # send errors from a no-reply address so we don't get into bounceback loops
+    no_reply_address = settings.NO_REPLY_ADDRESS
+
+    # short circuit if the mailing list doesn't exist
+    try:
+        ml = MailingList.objects.get_or_create_or_delete_mailing_list_by_address(
+                 recipient.address)
+    except MailingList.DoesNotExist:
+        logger.info(
+            u'Sending mailing list bounce back email to %s for mailing list %s '
+            u'because the mailing list does not exist', sender, recipient)
+        template = get_template('mailgun/email/bounce_back_does_not_exist.html')
+        content = template.render(Context({
+            'sender': sender,
+            'recipient': recipient.full_spec(),
+            'subject': subject,
+            'message_body': body_plain or body_html,
+        }))
+        listserv_client.send_mail(no_reply_address, no_reply_address, sender,
```

Comment:
Yes.  If we get an email addressed to two bogus mailing list addresses, we'll send bounces for both addresses.